src/python/pants/backend/rust/target_types.py

```diff
@@ -0,0 +1,30 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.target import COMMON_TARGET_FIELDS, MultipleSourcesField, Target, \
+    SingleSourceField
+
+
+class CargoTomlSourceField(SingleSourceField):
+    default = "Cargo.toml"
+
+
+class RustWorkspaceTarget(Target):
+    alias = "rust_workspace"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        CargoTomlSourceField,
+    )
+
+
+class RustPackageSourcesField(MultipleSourcesField):
+    default = ("Cargo.toml", "src/**/*.rs", "tests/**/*.rs")
```

Comment:
I think we should model this like the other plugins and just use nonrecurring globs.

Then the code that needs cargo.toml loads it as a config.

Reply:
Or keep the recursive globs, but do the cargo.toml loading as a config.

Reply:
For context, in a discussion I had @lassemand outside of this PR, the intent is, in a future PR, to have Cargo supply the content for `sources` automatically. My initial thought in that discussion had been to either remove the `sources` field (as unnecessary for Rust targets) or make it hidden if the need exists to pass through information from Cargo via targets to other rules like the `rustfmt` rules.

But we might still need a `sources` to allow adding sources which Cargo does not know about and which might be used by the Crate. In that case, the default here for sources would be empty.

We should probably document that plan as a TODO comment here.

Reply:
You'd want them to either be explicit or synthetic targets, for stuff like `--changed` to work

Reply:
As an example of our confusing terminology, it actually probably "generated targets" since those are targets which are created by an explicit target. Synthetic targets are conceptually also programmatically generated but are a different API.

In any event, something to consider in designing this support for a future PR.

Reply:
As I have limited background knowledge into the topics you are discussing here, I will have to ask whether we can work on this content as is, and resolve this conversation given a  corresponding TODO comment written here, as @tdyas suggested

Reply:
A TODO comment is sufficient in my view. Future PRs will continue to make improvements.

Reply:
I don't think we can leave this as a TODO. Sources get partitioned by file which means some partitions will have the toml file and some won't (leading to undesirable behavior)

Reply:
Shouldn't rustfmt be running in a single partition by default? If it isn't, then I agree that would be a problem.

Reply:
Sorry, nomenclature problem (we renamed things a while ago, so I'm still rewiring). It gets put into one big partition (thanks to the SINGLE_PARTITION) but that partition gets _batched_ into smaller groups (these used to be called partitions, hence the name problem)

That means one batch might have the Cargo toml while others won't.

Reply:
Ok. I will fix this issue as part of this PR. I will focus on fixing the skip test first. 

Realistically this issue will have to wait until next weekend.

Reply:
Absolutely no rush. If you want as well feel free to reach out and see if any of us in the community would be able and willing to give this effort a push by committing to your branch.

Reply:
This should leave out the cargo.toml as part of the partition. Lets include that at a later stage, as mentioned, as a config which can be given to the targets.